packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts

```diff
@@ -1090,6 +1094,202 @@ describe("SharedTree", () => {
+							revertible.retain();
+							undos.push(revertible);
+						}
+					},
+				);
+
+				if (edits === Edits.A) {
+					makeAPrerequisiteEdits(peer);
+				} else if (edits === Edits.B || edits === Edits.C) {
+					makeBAndCPrerequisiteEdits(peer);
+				}
+
+				unsubscribe();
+				return undos;
+			}
+
+			function makeAPrerequisiteEdits(
+				peer: CheckoutFlexTreeView<typeof schema.rootFieldSchema>,
+			): void {
+				const outerList = peer.flexTree.content.content;
+				const innerList = (outerList.at(0) ?? assert.fail()).content;
+				innerList.insertAtEnd("b");
+				innerList.insertAtEnd("c");
+			}
+
+			function makeBAndCPrerequisiteEdits(
+				peer: CheckoutFlexTreeView<typeof schema.rootFieldSchema>,
+			): void {
+				const outerList = peer.flexTree.content.content;
+				const innerList = (outerList.at(0) ?? assert.fail()).content;
+				innerList.insertAtEnd("d");
+				peer.flexTree.content.content.removeAt(0);
+			}
+
+			/**
+			 * Each scenario [Edits.X, Edits.Y] (where X and Y represent the possible values of Edits) describes two peers,
+			 * the first performs the edits described by Edits.X and the second performs the edits described by Edits.Y.
+			 * The edits performed by the second peer are sent through the resubmit code path.
+			 */
+			const scenarios = [
+				[Edits.A, Edits.A],
+				[Edits.A, Edits.B],
+				[Edits.A, Edits.C],
+				[Edits.B, Edits.A],
+				[Edits.C, Edits.A],
+			];
+
+			scenarios.forEach(([otherEdits, resubmitEdits]) => {
+				it(`${resubmitEdits} over ${otherEdits}`, () => {
+					const provider = new TestTreeProviderLite(2);
+					const content = {
+						schema,
+						allowedSchemaModifications: AllowedUpdateType.None,
+						initialTree: [["a"]],
+					};
+					const resubmitTree = provider.trees[0].schematizeInternal(content);
+					const otherTree = provider.trees[1].schematizeInternal(content);
+					provider.processMessages();
+
+					const otherUndos = makeEdits(otherTree, otherEdits);
+					const resubmitUndos = makeEdits(resubmitTree, resubmitEdits);
+
+					provider.processMessages();
+
+					const resubmitTreeOuterList = resubmitTree.flexTree.content.content;
+					const otherTreeOuterList = otherTree.flexTree.content.content;
+
+					if (otherEdits === Edits.A && resubmitEdits === Edits.A) {
+						assert.deepEqual(
+							// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+							[...resubmitTreeOuterList.at(0)!.content],
+							["a", "b", "c", "b", "c"],
+						);
+						assert.deepEqual(
+							// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+							[...otherTreeOuterList.at(0)!.content],
+							["a", "b", "c", "b", "c"],
+						);
+					} else {
+						assert.deepEqual([...resubmitTreeOuterList], []);
+						assert.deepEqual([...otherTreeOuterList], []);
+					}
+
+					assert.equal(otherUndos[1].revert(), RevertibleResult.Success);
+					assert.equal(otherUndos[0].revert(), RevertibleResult.Success);
```

Comment:
For Edits.EditAndRestore these need to be in the reverse order, right?

Reply:
yup, nice catch